### ISSUE TITLE
Multi-scale NMS ordering

### ISSUE THREAD

**Issue by asgersvenning** [darsa-group/flat-bug#91]
NMS (Non-Max-Suppression) is a greedy algorithm for deduplication of objects given a similarity metric, a duplicate similarity threshold. However, the ordering of the objects determine which object will be removed when two objects have a similarity above the threshold, and due to the greedy nature of the algorithm this can have non-local (i.e. cascading effects) that depend on the order, that is objects can potentially affect whether any other object is removed, no matter what their similarity is (for examples non-touching bounding boxes/polygons at any distance within a scene). It is therefore quite important that the order of the objects aligns with our expectations, usually people will simply use the confidence for instance/object ordering in MOD, however we have so far been using the product of the scale and the confidence instead, to decrease priority of instances seen at lower resolution. This is specified [here](https://github.com/darsa-group/flat-bug/blob/f0a89616dc5b7b1d4469a4838df740e7f1609ff0/src/flat_bug/predictor.py#L296), specifically on [line 312](https://github.com/darsa-group/flat-bug/blob/f0a89616dc5b7b1d4469a4838df740e7f1609ff0/src/flat_bug/predictor.py#L312) and [line 321](https://github.com/darsa-group/flat-bug/blob/f0a89616dc5b7b1d4469a4838df740e7f1609ff0/src/flat_bug/predictor.py#L321):
```
    def non_max_suppression(self, iou_threshold : float, **kwargs):
        """
        Simply wraps the nms_masks function from yolo_helpers.py, and removes the elements that were not selected.
        """
        if self.time:
            # Initialize timing calculations
            start = torch.cuda.Event(enable_timing=True)
            end = torch.cuda.Event(enable_timing=True)
            start.record()
            len_before = len(self)

        # Skip if there are no elements to merge
        if len(self) > 1:
            # Perform non-maximum suppression on the masks, using the scales as weights, that is the highest resolution masks are given the highest priority
            if self.PREFER_POLYGONS:
                nms_ind = nms_polygons(self.polygons,
HERE ==>>                           torch.tensor(self.scales, dtype=self.dtype, device=self.device) * self.confs,
                                    iou_threshold=iou_threshold, return_indices=True, dtype=self.dtype,
                                    boxes=self.boxes, **kwargs)
            else:
                image_to_mask_scale = torch.tensor(
                    [self.image.shape[1] / self.masks.data.shape[1], self.image.shape[2] / self.masks.data.shape[2]],
                    device=self.device, dtype=self.dtype
                )
                nms_ind = nms_masks(self.masks.data,
HERE ==>>                           torch.tensor(self.scales, dtype=self.dtype, device=self.device) * self.confs, # <<== HERE
                                    iou_threshold=iou_threshold, return_indices=True,
                                    boxes=self.boxes / image_to_mask_scale.repeat(2).unsqueeze(0), **kwargs)
            # Remove the elements that were not selected
            self = self[nms_ind]
```
I think this is a quite crucial design decision, that should receive further attention. 

I have a suspicion that it can cause issues like this: 
![image](https://github.com/darsa-group/flat-bug/assets/82270047/8999f2f9-654d-4e2a-81ab-9f531d4da8d9)
vs. this:
![image](https://github.com/darsa-group/flat-bug/assets/82270047/8f498198-3bc5-469d-8645-d86c6b658cad)
But if we were to fix this we should probably find a reproducible example where it matters a lot. For that purpose I think a scene with overlapping insects of different sizes would be good.
